File: 0581_Shortest_Unsorted_Continuous_Subarray/0581_Shortest_Unsorted_Continuous_Subarray.cc
#include <vector>
#include <algorithm>

class Solution {
public:
    int findUnsortedSubarray(std::vector<int>& nums) {
        std::vector<int> sorted_nums(nums.size());
        std::copy(nums.begin(), nums.end(), sorted_nums.begin());

        std::sort(sorted_nums.begin(), sorted_nums.end());
        if (nums == sorted_nums) {
            return 0;
        }

        int l = -1;
        int r = -1;
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            if (nums[i] != sorted_nums[i]) {
                l = i;
                break;
            }
        }

        for (int i = n-1; i >= 0; --i) {
            if (nums[i] != sorted_nums[i]) {
                r = i;
                break;
            }
        }

        return r-l+1;
    }
};

File: 0581_Shortest_Unsorted_Continuous_Subarray/0581_Shortest_Unsorted_Continuous_Subarray.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        N, sorted_nums = len(nums), sorted(nums)
        if all([nums[i] == sorted_nums[i] for i in range(0, N)]):
            return 0

        l, r = -1, -1
        for i in range(0, N):
            if nums[i] != sorted_nums[i]:
                l = i
                break

        for i in range(N-1, -1, -1):
            if nums[i] != sorted_nums[i]:
                r = i
                break

        return r-l+1